### PR TITLE
Add login logout

### PR DESCRIPTION
**Pull request by msm-faraj**
This pr aims to:

- add authHandle middleware
- add login method to authController
- add logout method to authController 
- add genereateToken method to the User mode

### CODE REVIEW

File: src/config.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 module.exports = {
   authentication: {
     salt: 10,
+    jwePrivateKey: "stockJournalJWTPrivateKey",
```
Comment (pouyapourbaba) [inline src/config.js:4]:
```suggestion
    jwtPrivateKey: "stockJournalJWTPrivateKey",
```

File: src/controllers/authController.js [hunk 2]
```diff
@@ -27,6 +27,43 @@ class AuthController {
...
+  async logIn(req, res) {
+    let { email, password } = req.body;
+    let user = await this.User.findOne({
+      where: { email },
+    });
+    // check user existance
+    if (!user)
+      return res
+        .status(httpStatus.BAD_REQUEST)
+        .send("Invalid email of password.");
```
Comment (pouyapourbaba) [inline src/controllers/authController.js:40]:
```suggestion
        .send("Invalid email or password.");
```

File: src/controllers/authController.js [hunk 3]
```diff
@@ -27,6 +27,43 @@ class AuthController {
...
+
+    // check password
+    const validPassword = await bcrypt.compare(password, user.password);
```
Comment (pouyapourbaba) [inline src/controllers/authController.js:43]:
```suggestion
    const isPasswordMatch = await bcrypt.compare(password, user.password);
```

File: src/controllers/authController.js [hunk 4]
```diff
@@ -27,6 +27,43 @@ class AuthController {
...
+
+    // create jwt token
+    const token = user.generateToken();
+
+    // wrtie jwt to the database
+    user.jwt = token;
+    await user.save();
+
+    // set token to the req.header
+    res.header("x-auth-token", token).send(token);
```
Comment (pouyapourbaba) [inline src/controllers/authController.js:57]: Put `x-auth-token` into a constant, you can call it `AUTH_HEADER_NAME`

File: src/controllers/authSchema.js [hunk 5]
```diff
@@ -1,9 +1,14 @@
...
-const registerSchema = Joi.object({
+module.exports.registerSchema = Joi.object({
   username: Joi.string().min(3).required(),
   password: Joi.number().min(1).required(),
   email: Joi.string().email().required(),
 });
 
-module.exports = registerSchema;
+module.exports.loginSchema = Joi.object({
+  password: Joi.number().min(1).required(),
```
Comment (pouyapourbaba) [inline src/controllers/authSchema.js:10]:
```suggestion
  password: Joi.string().min(1).required(),
```

File: src/controllers/authSchema.js [hunk 6]
```diff
@@ -1,9 +1,14 @@
 const Joi = require("joi");
 
-const registerSchema = Joi.object({
+module.exports.registerSchema = Joi.object({
   username: Joi.string().min(3).required(),
   password: Joi.number().min(1).required(),
```
Comment (pouyapourbaba) [inline src/controllers/authSchema.js:5]: Password is gonna be alphanumeric

File: src/middlewares/auth-handler.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+      .send("Access denied. No token provided.");
+  try {
+    const decodedPayload = jwt.verify(
+      receivedToken,
+      config.authentication.jwePrivateKey
+    );
+    req.authenticatedUser = decodedPayload; // {id: "user.id"}
+    next();
+  } catch (error) {
+    res.status(httpStatus.BAD_REQUEST).send("Invalid token.");
```
Comment (pouyapourbaba) [inline src/middlewares/auth-handler.js:19]:
```suggestion
    res.status(httpStatus.UNAUTHORIZED).send("Invalid token.");
```

File: src/models/index.js [hunk 8]
```diff
@@ -1,5 +1,5 @@
-module.exports = (sequelize, dataTypes) => {
-  const User = require("./user.js")(sequelize, dataTypes);
+module.exports = (sequelize, dataTypes, jwt, config) => {
```
Comment (pouyapourbaba) [inline src/models/index.js:1]: Let's keep this clean, remove jwt and config from the function arguments. Require these two in the User model file itself.

File: src/controllers/authController.js [hunk 9]
```diff
@@ -27,6 +27,43 @@ class AuthController {
...
+  }
+
+  async logOut(req, res) {
```
Comment (pouyapourbaba) [inline src/controllers/authController.js:60]: I believe you can update the user without finding it here, we do not care if the user exists or not. 
```
await User.update({jwt: null}, {where: {id: userId}})
```

And always respond to the request.

File: src/config.js (hunk 1 above)
Comment (msm-faraj) [reply src/config.js:4]: done

File: src/controllers/authController.js (hunk 2 above)
Comment (msm-faraj) [reply src/controllers/authController.js:40]: done

File: src/controllers/authController.js (hunk 3 above)
Comment (msm-faraj) [reply src/controllers/authController.js:43]: done

File: src/controllers/authController.js (hunk 4 above)
Comment (msm-faraj) [reply src/controllers/authController.js:57]: done

File: src/controllers/authSchema.js (hunk 5 above)
Comment (msm-faraj) [reply src/controllers/authSchema.js:10]: done

File: src/middlewares/auth-handler.js (hunk 7 above)
Comment (msm-faraj) [reply src/middlewares/auth-handler.js:19]: done

File: src/models/index.js (hunk 8 above)
Comment (msm-faraj) [reply src/models/index.js:1]: done

File: src/controllers/authController.js (hunk 9 above)
Comment (msm-faraj) [reply src/controllers/authController.js:60]: done

File: src/controllers/authSchema.js (hunk 6 above)
Comment (msm-faraj) [reply src/controllers/authSchema.js:5]: done

**Review by pouyapourbaba — APPROVED** [review]

**Pull request merged**